File: python/video_sampling.py
"""
Adaptive frame sampling for Video RLVO (Python side).

Reads the SAME shared/video-sampling.json as src/lib/videoSampling.ts. The frame counts used to be
duplicated — SUMMARY_FRAMES/TIMECAPSULE_FRAMES here, separate literals in the React page — so
changing one silently left the other behind. There is now one place to change them.

Consistency with the browser
----------------------------
The browser seeks by TIME (`duration`), this module selects by FRAME INDEX (`CAP_PROP_FRAME_COUNT`).
Both sample the same fractional positions i/(n-1) across the video, so for constant-framerate
material they land on the same moments. They can differ on variable-framerate recordings, where a
frame index is not proportional to elapsed time — an inherent property of the two APIs, not a
disagreement about policy.
"""

import json
from pathlib import Path

_CONFIG_PATH = Path(__file__).resolve().parent.parent / "shared" / "video-sampling.json"

with _CONFIG_PATH.open(encoding="utf-8") as _f:
    _CONFIG = json.load(_f)

TIERS = _CONFIG["tiers"]
END_EPSILON_SECONDS = float(_CONFIG["endEpsilonSeconds"])

MODES = tuple(TIERS.keys())


def frame_count_for(mode: str, duration_seconds: float) -> int:
    """Frames to extract for a video of this length, per the shared tiers."""
    tiers = TIERS.get(mode)
    if not tiers:
        raise ValueError(f'No sampling tiers configured for mode "{mode}". Known: {", ".join(MODES)}')

    # An unknown or zero duration falls back to the shortest tier: fewer frames still produces a
    # usable result, where raising would lose the whole request.
    if not duration_seconds or duration_seconds <= 0:
        return int(tiers[0]["frames"])

    for tier in tiers:
        cap = tier["maxSeconds"]
        if cap is None or duration_seconds <= cap:
            return int(tier["frames"])
    return int(tiers[-1]["frames"])


def frame_indices_for(total_frames: int, count: int) -> list:
    """
    Evenly spaced frame indices, first at 0 and last at the final frame.

    Dividing by `count - 1` rather than `count` is what makes the last sample reach the end of the
    video. The previous `i * (total // count)` stepping stopped short, so the closing portion of
    every video went unsampled.
    """
    if count <= 0 or total_frames <= 0:
        return []
    last = total_frames - 1
    if count == 1:
        return [0]
    return [min(round(i * last / (count - 1)), last) for i in range(count)]


def sample_timestamps(duration_seconds: float, count: int) -> list:
    """Time-based equivalent of frame_indices_for — mirrors the browser exactly."""
    if count <= 0:
        return []
    duration = duration_seconds if duration_seconds and duration_seconds > 0 else 0.0
    end = max(0.0, duration - END_EPSILON_SECONDS)
    if count == 1:
        return [0.0]
    return [min(i * duration / (count - 1), end) for i in range(count)]

File: src/lib/videoSampling.ts
/**
 * Adaptive frame sampling for Video RLVO.
 *
 * WHAT CHANGED AND WHY
 * Frame counts were fixed literals — 6 for summary, 8 for time capsule — so a five-second clip and
 * a two-hour film were both reduced to the same handful of frames. Short clips paid for redundant
 * near-identical frames; long ones lost almost everything between samples. The count now follows
 * the video's duration.
 *
 * The tiers live in shared/video-sampling.json and are read by python/video_sampling.py as well, so
 * the two implementations cannot drift. They previously each carried their own copy of the numbers.
 *
 * THE END-OF-VIDEO FIX
 * The old loop spaced samples by `duration / n` and stepped `i = 0 … n-1`, so the last sample landed
 * at `(n-1)/n` of the way through and the closing stretch of every video was never looked at — for
 * summary, the final sixth. Spacing over `n - 1` intervals instead puts the first sample at 0 and
 * the last at the end, which is what "evenly spaced across the video" should have meant.
 */
import config from "../../shared/video-sampling.json";

export type SamplingMode = "summary" | "timecapsule";

interface Tier {
  maxSeconds: number | null;
  frames: number;
}

const TIERS = config.tiers as Record<SamplingMode, Tier[]>;
const END_EPSILON = config.endEpsilonSeconds as number;

/**
 * How many frames to take for a video of this length.
 *
 * A non-finite duration (a stream, or metadata that never loaded) falls back to the shortest tier
 * rather than throwing: fewer frames still produces a usable result, whereas a thrown error loses
 * the upload entirely.
 */
export function frameCountFor(mode: SamplingMode, durationSeconds: number): number {
  const tiers = TIERS[mode];
  if (!tiers?.length) throw new Error(`No sampling tiers configured for mode "${mode}".`);

  if (!Number.isFinite(durationSeconds) || durationSeconds <= 0) return tiers[0].frames;

  const tier = tiers.find((t) => t.maxSeconds === null || durationSeconds <= t.maxSeconds);
  return (tier ?? tiers[tiers.length - 1]).frames;
}

/**
 * Evenly spaced timestamps covering the whole video, first at 0 and last at the end.
 *
 * Spacing divides by `count - 1`, not `count`, which is what makes the final sample reach the end.
 * The last target is pulled back by END_EPSILON because seeking to exactly `duration` commonly
 * decodes nothing.
 */
export function sampleTimestamps(durationSeconds: number, count: number): number[] {
  if (count <= 0) return [];
  const duration = Number.isFinite(durationSeconds) && durationSeconds > 0 ? durationSeconds : 0;
  const end = Math.max(0, duration - END_EPSILON);
  if (count === 1) return [0];

  return Array.from({ length: count }, (_, i) => Math.min((i * duration) / (count - 1), end));
}

/** Convenience: the timestamps this mode would use for a video of this length. */
export function timestampsFor(mode: SamplingMode, durationSeconds: number): number[] {
  return sampleTimestamps(durationSeconds, frameCountFor(mode, durationSeconds));
}
